src/main/java/org/privacyidea/JSONParser.java

```diff
@@ -0,0 +1,345 @@
+
+            JsonElement errElem = result.get(ERROR);
+            if (errElem != null && !errElem.isJsonNull()) {
+                JsonObject errObj = result.getAsJsonObject(ERROR);
+                response.error = new PIResponse.Error();
+                response.error.code = getInt(errObj, CODE);
+                response.error.message = getString(errObj, MESSAGE);
+            }
+        }
+
+        JsonElement detailElem = obj.get(DETAIL);
+        if (detailElem != null && !detailElem.isJsonNull()) {
+            JsonObject detail = obj.getAsJsonObject(DETAIL);
+            response.message = getString(detail, MESSAGE);
+            response.serial = getString(detail, SERIAL);
+            response.transactionID = getString(detail, TRANSACTION_ID);
+            response.type = getString(detail, TYPE);
+            response.otpLength = getInt(detail, OTPLEN);
+
+            JsonArray arrMessages = detail.getAsJsonArray(MESSAGES);
+            if (arrMessages != null) {
+                arrMessages.forEach(val -> {
+                    if (val != null) {
+                        response.messages.add(val.getAsString());
+                    }
+                });
+            }
+
+            JsonArray arrChallenges = detail.getAsJsonArray(MULTI_CHALLENGE);
+            if (arrChallenges != null) {
+                for (int i = 0; i < arrChallenges.size(); i++) {
+                    JsonObject challenge = arrChallenges.get(i).getAsJsonObject();
+                    if (TOKEN_TYPE_WEBAUTHN.equals(getString(challenge, TYPE))) {
+                        String webAuthnSignRequest = "";
+                        JsonElement attrElem = challenge.get(ATTRIBUTES);
+                        if (attrElem != null && !attrElem.isJsonNull()) {
+                            JsonElement webauthnElem = attrElem.getAsJsonObject().get(WEBAUTHN_SIGN_REQUEST);
+                            if (webauthnElem != null && !webauthnElem.isJsonNull()) {
+                                webAuthnSignRequest = webauthnElem.toString();
+                            }
+                        }
+                        response.multichallenge.add(new WebAuthn(
+                                getString(challenge, SERIAL),
+                                getString(challenge, MESSAGE),
+                                getString(challenge, TRANSACTION_ID),
+                                webAuthnSignRequest
+                        ));
+                    } else {
+                        response.multichallenge.add(new Challenge(
+                                getString(challenge, SERIAL),
+                                getString(challenge, MESSAGE),
+                                getString(challenge, TRANSACTION_ID),
+                                getString(challenge, TYPE)
+                        ));
+                    }
+                }
+            }
+        }
+        return response;
+    }
+
+    List<TokenInfo> parseTokenInfo(String serverResponse) {
```

Comment:
Are you parsing and returning tokenobjects or tokeninfo?

In the server naming contect the "tokeninfo" are only additional, freely choosable token attributes.

Reply:
this parses the response of GET /token (which is done to check if a token should be enrolled). I named it TokenInfo so there is no misconception if the name was Token that it is a real software token of some kind.
TokenInfo also contains the info map.